framework/include/materials/Material.h

```diff
@@ -298,3 +317,25 @@ Material::getMaterialPropertyOlderByName(const std::string & prop_name_in)
   registerPropName(prop_name, true, Material::OLDER);
   return MaterialPropertyInterface::getMaterialPropertyOlderByName<T>(prop_name_in);
 }
+
+#include "OptionalMaterialPropertyProxy.h"
+
+template <typename T>
+const MaterialProperty<T> * const &
+Material::getOptionalMaterialProperty(const std::string & name)
+{
+  auto proxy = std::make_unique<OptionalMaterialPropertyProxy<T>>(name, this);
+  auto pointer = proxy->value();
+  _fe_problem.addOptionalMaterialPropertyProxy(proxy);
```

Comment:
Ugh I really hate having `FEProblemBase` be the storage location for all of our garbage.

Can we not keep track of these on each `Material`? If we ever are actually able to do dependency based execution, keeping stuff like this local to the relevant object would be nice.

Not a requirement, just a thought

But, if you keep it like this - I'm pretty sure this leaks memory. You build a unique pointer, pass a ref of it to the problem, and then it goes out of scope

Reply:
it doesn't leak because it doesn't even compile :-D

But the idea was to pass as non const ref and move into the vector of unique pointers. When `proxy` goes out of scope it's already invalid.

I had to change it to pass a raw pointer and emplace into the vector of unique ptrs for simple handling of teh polymorphism.